File: _3_b_algorithms_in_python/_1_recursion_and_backtracking/_1_basics/_8_power_of_number.py
#  https://www.geeksforgeeks.org/write-a-c-program-to-calculate-powxn/
#  https://leetcode.com/problems/powx-n/

#  brute force using recursion
def power(x, n):
    if n == 0:
        return 1
    return power(x, n - 1) * x


#  brute force iterative
def power(x, n):
    ans = 1.0
    for i in range(n):
        ans *= x
    return ans
#  Time Complexity: O(N)
#  Space Complexity: O(1)

#  using binary exponentiation
def power(x, n):
    ans = 1.0
    nn = abs(n)

    while nn:
        if nn % 2:
            ans *= x
            nn -= 1
        else:
            x *= x
            nn //= 2

    if n < 0:
        ans = 1.0 / ans

    return ans

# Example usage:
result = power(2.0, 10)
print(result)

#  Time Complexity: O(log n)
#  Space Complexity: O(1)

File: _10_backtracking/_1_basics/_8_power_of_number.cpp
// https://www.geeksforgeeks.org/write-a-c-program-to-calculate-powxn/
// https://leetcode.com/problems/powx-n/
#include <bits/stdc++.h>
using namespace std;

// brute force using recursion
int power(int x, int n)
{
    if (n == 0)
        return 1;
    return power(x, n - 1) * x;
}

// brute force iterative
double power(double x, int n)
{
    double ans = 1.0;
    for (int i = 0; i < n; i++)
    {
        ans = ans * x;
    }
    return ans;
}
// Time Complexity: O(N)
// Space Complexity: O(1)

// using binary exponentiation
double power(double x, int n)
{
    double ans = 1.0;
    long long nn = n;
    if (nn < 0)
        nn = -1 * nn;
    while (nn)
    {
        if (nn % 2)
        {
            ans = ans * x;
            nn = nn - 1;
        }
        else
        {
            x = x * x;
            nn = nn / 2;
        }
    }
    if (n < 0)
        ans = (double)(1.0) / (double)(ans);
    return ans;
}
// Time Complexity: O(log n)
// Space Complexity: O(1)

int main()
{
    int x, n;
    cin >> x >> n;
    cout << power(x, n);
}